[Coverage] Avoid recording regions from macro expansions

hamishknight opened the pull request:
Ignore any regions recorded while inside a macro expansion, but account for any control flow that may have happened such that the exit count is correctly adjusted. This allows e.g throwing function calls to behave correctly within the expansion.

Additionally, make sure we avoid profiling functions with function body macros.

rdar://129081384

hamishknight commented:
@swift-ci please test

hamishknight commented on test/Profiler/coverage_macros.swift:
```diff
@@ -43,7 +70,62 @@ struct S2 {
+  guard .random() else { return } // CHECK-NEXT: [[@LINE]]:24   -> [[@LINE]]:34  : 1
+                                  // CHECK-NEXT: [[@LINE-1]]:34 -> [[@LINE+2]]:2 : (0 - 1)
+  #declMacroWithControlFlow       // CHECK-NEXT: [[@LINE]]:28   -> [[@LINE+1]]:2 : ((0 - 1) - 3)
+}
```
Need to also add a case for a nested expansion

hamishknight commented:
@swift-ci please test

ahoppen commented on lib/SIL/IR/SILProfiler.cpp:
```diff
@@ -247,9 +258,10 @@ bool shouldSkipExpr(Expr *E) {
+  // We want to walk into initializers for bindings, and the expansions of
+  // MacroExpansionDecls, which will be nested within MacroExpansionExprs in
+  // local contexts.
+  return isa<PatternBindingDecl>(D) || isa<MacroExpansionDecl>(D);
```
How do we report code coverage for the expansions of a macro? Or I don’t fully understand why we need to walk into `MacroExpansionDecl`.

hamishknight replied:
We don't, the only reason we need to walk into the expansion is to get accurate counter information for the following code, e.g for `#id(try throwingFn())`, we need to account for the fact that code after the macro may not run if an error was thrown

ahoppen reviewed (approved):
Looks reasonable to me. But I’m not too familiar with code coverage overall.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight committed:
[Coverage] Also walk MacroExpansionDecls

Make sure we walk macro expansion decls to handle
cases where e.g a binding introduces control flow.
This should be a pretty uncommon case since bindings
introduced by macros aren't actually usable when
expanded in local contexts, but handle it all the
same.

hamishknight committed:
[Coverage] Avoid profiling functions with body macros

These replace the body of the function, and as such
shouldn't be profiled.

hamishknight committed:
[Coverage] Avoid recording regions from macro expansions

Ignore any regions recorded while inside a macro
expansion, but account for any control flow that
may have happened such that the exit count is
correctly adjusted. This allows e.g throwing function
calls to behave correctly within the expansion.

rdar://129081384

hamishknight committed:
[Coverage] NFC: Set the SourceFile on CoverageMapping

Saves having to pass it as a parameter to
`emitSourceRegions`.

hamishknight commented:
@swift-ci please test
